test/integration/auth/auth.prose.test.ts

```diff
@@ -0,0 +1,373 @@
+          await client.close();
+        }
+      });
+
+      /*
+       * Step 3
+       * For a non-existent username, verify that not specifying a mechanism when
+       * connecting fails with the same error type that would occur with a correct
+       * username but incorrect password or mechanism.  (Because negotiation with a
+       * non-existent user name at one point during server development caused a
+       * handshake error, we want to verify this is seen by users as similar to other
+       * authentication errors, not as a network or database command error on the ``hello``
+       * or legacy hello commands themselves.)
+       */
+      it('fails for a nonexistent username with same error type as bad password', {
+        metadata: { requires: { mongodb: '>=3.7.3' } },
+        test: async function () {
+          const noUsernameOptions = {
+            auth: {
+              username: 'roth',
+              password: 'pencil'
+            },
+            authSource: 'admin'
+          };
+
+          const badPasswordOptions = {
+            auth: {
+              username: 'both',
+              password: 'pencil'
+            },
+            authSource: 'admin'
+          };
+
+          const noUserClient = this.configuration.newClient({}, noUsernameOptions);
+          const badPasswordClient = this.configuration.newClient({}, badPasswordOptions);
+          const noUserError = await noUserClient
+            .db('test')
+            .stats()
+            .catch(e => e);
+          const badPasswordError = await badPasswordClient
+            .db('test')
+            .stats()
+            .catch(e => e);
+          expect(noUserError).to.match(/Authentication failed/);
+          expect(badPasswordError).to.match(/Authentication failed/);
+          await noUserClient.close();
+          await badPasswordClient.close();
+        }
+      });
+
+      it('should send speculativeAuthenticate on initial handshake on MongoDB 4.4+', {
+        metadata: { requires: { mongodb: '>=4.4', topology: ['single'] } },
+        test: async function () {
+          const options = {
+            auth: {
+              username: userMap.both.username,
+              password: userMap.both.password
+            },
+            authSource: 'admin'
+          };
+
+          const commandSpy = sinon.spy(Connection.prototype, 'command');
+          const client = this.configuration.newClient({}, options);
+          await client.connect();
+          const calls = commandSpy
+            .getCalls()
+            .filter(c => c.thisValue.id !== '<monitor>') // ignore all monitor connections
+            .filter(
+              c => c.args[1][process.env.MONGODB_API_VERSION ? 'hello' : LEGACY_HELLO_COMMAND]
+            );
+
+          expect(calls).to.have.length(1);
+          const handshakeDoc = calls[0].args[1];
+          expect(handshakeDoc).to.have.property('speculativeAuthenticate');
+          sinon.restore();
```

Comment:
same comment about leaking clients with the `sinon.restore()`

Reply:
Use try...finally now.